### PR TITLE
Strafe

### PR DESCRIPTION
**Pull request by ChrisCGalbraith**
Implemented logic if a Shred or Backstab fails to cast, will strafe until behind the target and a successful cast lands.

### CODE REVIEW

File: MaloWBotLK/MaloWBotLK.lua [hunk 1]
```diff
@@ -101,6 +117,8 @@ f:RegisterEvent("READY_CHECK")
 f:RegisterEvent("UNIT_SPELLCAST_START")
 f:RegisterEvent("UNIT_SPELLCAST_CHANNEL_START")
 f:RegisterEvent("UNIT_TARGET")
+f:RegisterEvent("UNIT_SPELLCAST_FAILED")
+f:RegisterEvent("UNIT_SPELLCAST_SUCCEEDED")
```
Comment (Malow) [inline MaloWBotLK/MaloWBotLK.lua]: I think using UNIT_SPELLCAST_FAILED/UNIT_SPELLCAST_SUCCEEDED will lead to some situations that is undesirable, for example on Kel'Thuzad when they strafe out of void zones they will stop the strafing if they successfully land a backstab or shred. Also if they try to cast Shred or Backstab and don't have enough energy they will get a UNIT_SPELLCAST_FAILED which will cause them to start strafing even though they're in the right position. I think using the error text "You must be behind the target" is a better hook.

File: MaloWBotLK/MaloWBotLK.lua [hunk 2]
```diff
@@ -203,9 +221,9 @@ function mb_UpdateClassOrder()
 	local name = UnitName("player")
 	mb_classOrder = {}
 	table.insert(mb_classOrder, name)
-    local members = mb_GetNumPartyOrRaidMembers()
-    for i = 1, members do
-        local unit = mb_GetUnitFromPartyOrRaidIndex(i)
+	local members = mb_GetNumPartyOrRaidMembers()
+	for i = 1, members do
+		local unit = mb_GetUnitFromPartyOrRaidIndex(i)
```
Comment (Malow) [inline MaloWBotLK/MaloWBotLK.lua]: You have a lot of indentation changes like these that should not be part of this PR

File: MaloWBotLK/MaloWBotLK.lua [hunk 3]
```diff
@@ -83,6 +83,22 @@ function mb_OnEvent(self, event, arg1, arg2, arg3, arg4, ...)
 		if arg1 == "player" and mb_registeredInterruptSpells ~= nil then
 			mb_HandleTargetSpellcast()
 		end
+	elseif event == "UNIT_SPELLCAST_FAILED" then
+		if arg1 == "player" and (arg2 == "Shred" or arg2 == "Backstab") then
+			mb_lastStrafe = mb_time
+			if mb_lastStrafe + 0.5 > mb_time then
```
Comment (Malow) [inline MaloWBotLK/MaloWBotLK.lua]: This code seems quite dangerous, if a strafe is started there are multiple scenarios where the strafe won't ever be cancelled. I would rework this logic to start a strafe using StrafeRightStart and at the same time setting an mb_shouldStopStrafingAt variable to mb_time + 0.1 when it's detected that you're not behind your target, and then in mb_HandleAutomaticMovement I would check mb_shouldStopStrafingAt compared to mb_time and stop the strafe if that 0.1 sec has passed, similar to how mb_shouldStopMovingForwardAt currently works.

**Review by Malow — CHANGES_REQUESTED** [review]

File: MaloWBotLK/MaloWBotLK.lua (hunk 2 above)
Comment (ChrisCGalbraith) [reply MaloWBotLK/MaloWBotLK.lua]: Reverted indentation changes.

File: MaloWBotLK/MaloWBotLK.lua (hunk 3 above)
Comment (ChrisCGalbraith) [reply MaloWBotLK/MaloWBotLK.lua]: eee65c9 : Now uses "You must be behind your target. " UI_ERROR message in order to start strafing.

File: MaloWBotLK/MaloWBotLK.lua (hunk 1 above)
Comment (ChrisCGalbraith) [reply MaloWBotLK/MaloWBotLK.lua]: Removed these checks now.

File: MaloWBotLK/MaloWBotLK.lua [hunk 4]
```diff
@@ -101,6 +106,8 @@ f:RegisterEvent("READY_CHECK")
 f:RegisterEvent("UNIT_SPELLCAST_START")
 f:RegisterEvent("UNIT_SPELLCAST_CHANNEL_START")
 f:RegisterEvent("UNIT_TARGET")
+f:RegisterEvent("UNIT_SPELLCAST_FAILED")
+f:RegisterEvent("UNIT_SPELLCAST_SUCCEEDED")
```
Comment (Malow) [inline MaloWBotLK/MaloWBotLK.lua]: These are not needed anymore right?

**Review by Malow — CHANGES_REQUESTED** [review]

File: MaloWBotLK/MaloWBotLK.lua (hunk 4 above)
Comment (ChrisCGalbraith) [reply MaloWBotLK/MaloWBotLK.lua]: Removed.

**Review by Malow — APPROVED** [review]
Looks good! Gj!
